File: packages/domain/src/shell-config.ts
import { z } from "zod";

import {
  ProviderTypeSchema,
  TechnicianCertificateKindSchema,
  TechnicianVerifiedLevelSchema,
  UserStatusSchema,
} from "./user";

// ───────── Provider mode ─────────

export const ProviderModeSchema = z.enum(["business", "individual"]);
export type ProviderMode = z.infer<typeof ProviderModeSchema>;

// ───────── Home layout variants ─────────

export const HomeLayoutSchema = z.enum([
  "tow_focused",
  "full",
  "business_lite",
  "minimal",
  "damage_shop",
]);
export type HomeLayout = z.infer<typeof HomeLayoutSchema>;

// ───────── Quick action ─────────

export const QuickActionSchema = z.object({
  id: z.string(),
  label: z.string(),
  icon: z.string(),
  route: z.string(),
  requires_capability: z.string().nullable().default(null),
});
export type QuickAction = z.infer<typeof QuickActionSchema>;

// ───────── Shell config ─────────

export const ShellConfigSchema = z.object({
  primary_provider_type: ProviderTypeSchema,
  active_provider_type: ProviderTypeSchema,
  provider_mode: ProviderModeSchema,
  secondary_provider_types: z.array(ProviderTypeSchema).default([]),
  verified_level: TechnicianVerifiedLevelSchema,
  admission_status: UserStatusSchema,
  admission_gate_passed: z.boolean().default(false),
  enabled_capabilities: z.array(z.string()).default([]),
  home_layout: HomeLayoutSchema,
  tab_set: z.array(z.string()),
  quick_action_set: z.array(QuickActionSchema).default([]),
  required_onboarding_steps: z.array(z.string()).default([]),
  required_cert_kinds: z.array(TechnicianCertificateKindSchema).default([]),
  role_config_version: z.number().int(),
});
export type ShellConfig = z.infer<typeof ShellConfigSchema>;

// ───────── V1 tab slot constants ─────────

export const V1_FIXED_TAB_SET = ["home", "havuz", "kayitlar", "profil"] as const;
export type V1FixedTabId = (typeof V1_FIXED_TAB_SET)[number];

File: naro-backend/app/schemas/shell_config.py
"""ShellConfig Pydantic — packages/domain/src/shell-config.ts parity.

Mobil service-app home/hava/kayitlar/profil shell'ini backend'den alır. Cache
key: `shell_config:{user_id}:{role_config_version}`; TTL 300s. Her mutation
`bump_role_config_version` çağırır → yeni versiyon → eski cache natural stale.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from app.models.technician import (
    ProviderMode,
    ProviderType,
    TechnicianCertificateKind,
    TechnicianVerifiedLevel,
)
from app.models.user import UserStatus


class HomeLayout(StrEnum):
    TOW_FOCUSED = "tow_focused"
    FULL = "full"
    BUSINESS_LITE = "business_lite"
    MINIMAL = "minimal"
    DAMAGE_SHOP = "damage_shop"


class QuickAction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    label: str
    icon: str
    route: str
    requires_capability: str | None = None


class ShellConfig(BaseModel):
    """V1 shell config — Zod canonical parity."""

    model_config = ConfigDict(extra="forbid")

    primary_provider_type: ProviderType
    active_provider_type: ProviderType
    provider_mode: ProviderMode
    secondary_provider_types: list[ProviderType] = Field(default_factory=list)
    verified_level: TechnicianVerifiedLevel
    admission_status: UserStatus
    admission_gate_passed: bool = False
    enabled_capabilities: list[str] = Field(default_factory=list)
    home_layout: HomeLayout
    tab_set: list[str]
    quick_action_set: list[QuickAction] = Field(default_factory=list)
    required_onboarding_steps: list[str] = Field(default_factory=list)
    required_cert_kinds: list[TechnicianCertificateKind] = Field(default_factory=list)
    role_config_version: int


# V1 fixed tab set — Zod canonical ile aynı
V1_FIXED_TAB_SET: list[str] = ["home", "havuz", "kayitlar", "profil"]
